<a href="https://colab.research.google.com/github/DJohanTwT/DataV24/blob/main/C2_Clase_5_Dataframe.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **0. Libraries**

In [1]:
import numpy as np          # Algebraic operations
import pandas as pd         # Manage the data base
import plotly.express as px # Graphics

# **1. Import the dataset**

In [2]:
# Save the file path
p = '/content/data.xlsx'

# Read and save the dataset
d = pd.read_excel(p)

# **2. Basic Dataframe exploration**

In [3]:
# General info
d.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1519 entries, 0 to 1518
Data columns (total 10 columns):
 #   Column                                                                    Non-Null Count  Dtype 
---  ------                                                                    --------------  ----- 
 0   Country Name                                                              1519 non-null   object
 1   Country Code                                                              1519 non-null   object
 2   Time                                                                      1519 non-null   int64 
 3   Time Code                                                                 1519 non-null   object
 4   Exports of goods and services (% of GDP) [NE.EXP.GNFS.ZS]                 1519 non-null   object
 5   Imports of goods and services (% of GDP) [NE.IMP.GNFS.ZS]                 1519 non-null   object
 6   Foreign direct investment, net inflows (% of GDP) [BX.KLT.DINV.WD.GD.ZS]

In [5]:
# View the first 5 rows
d.head()

Country Name Country Code  Time Time Code  \
0  Afghanistan          AFG  2017    YR2017   
1  Afghanistan          AFG  2018    YR2018   
2  Afghanistan          AFG  2019    YR2019   
3  Afghanistan          AFG  2020    YR2020   
4  Afghanistan          AFG  2021    YR2021   

  Exports of goods and services (% of GDP) [NE.EXP.GNFS.ZS]  \
0                                                 ..          
1                                                 ..          
2                                                 ..          
3                                          10.420817          
4                                          14.342153          

  Imports of goods and services (% of GDP) [NE.IMP.GNFS.ZS]  \
0                                                 ..          
1                                                 ..          
2                                                 ..          
3                                          36.289077          
4                                          37.069564          

  Foreign direct investment, net inflows (% of GDP) [BX.KLT.DINV.WD.GD.ZS]  \
0                                           0.274797                         
1                                           0.661572                         
2                                           0.124496                         
3                                           0.064994                         
4                                           0.144401                         

  Trade (% of GDP) [NE.TRD.GNFS.ZS] GDP growth (annual %) [NY.GDP.MKTP.KD.ZG]  \
0                                ..                                  2.647003   
1                                ..                                  1.189228   
2                                ..                                  3.911603   
3                         46.709895                                 -2.351101   
4                         51.411716                                -20.738839   

  External balance on goods and services (% of GDP) [NE.RSB.GNFS.ZS]  
0                                                 ..                  
1                                                 ..                  
2                                                 ..                  
3                                          -25.86826                  
4                                         -22.727411

In [4]:
# View the last 5 rows
d.tail()

Country Name Country Code  Time Time Code  \
1514     Zimbabwe          ZWE  2019    YR2019   
1515     Zimbabwe          ZWE  2020    YR2020   
1516     Zimbabwe          ZWE  2021    YR2021   
1517     Zimbabwe          ZWE  2022    YR2022   
1518     Zimbabwe          ZWE  2023    YR2023   

     Exports of goods and services (% of GDP) [NE.EXP.GNFS.ZS]  \
1514                                          27.554156          
1515                                          22.293063          
1516                                          22.775241          
1517                                          27.955246          
1518                                                 ..          

     Imports of goods and services (% of GDP) [NE.IMP.GNFS.ZS]  \
1514                                          28.241805          
1515                                          25.020302          
1516                                          28.071888          
1517                                          37.001395          
1518                                                 ..          

     Foreign direct investment, net inflows (% of GDP) [BX.KLT.DINV.WD.GD.ZS]  \
1514                                           1.142806                         
1515                                           0.699034                         
1516                                           0.881174                         
1517                                            1.24787                         
1518                                                 ..                         

     Trade (% of GDP) [NE.TRD.GNFS.ZS]  \
1514                         55.795961   
1515                         47.313365   
1516                         50.847128   
1517                         64.956641   
1518                                ..   

     GDP growth (annual %) [NY.GDP.MKTP.KD.ZG]  \
1514                                 -6.332446   
1515                                 -7.816951   
1516                                  8.468017   
1517                                  6.522375   
1518                                  4.955782   

     External balance on goods and services (% of GDP) [NE.RSB.GNFS.ZS]  
1514                                          -0.687648                  
1515                                          -2.727239                  
1516                                          -5.296647                  
1517                                          -9.046149                  
1518                                                 ..

# **3. Cleaning**

## 3.1. Column names

In [6]:
# Time to year
d.rename(columns={'Time': 'year'}, inplace=True)

# Country name to country
d.rename(columns={'Country Name': 'country'}, inplace=True)

# Country code to code
d.rename(columns={'Country Code': 'code'}, inplace=True)

# Exports of goods and services (% of GDP) [NE.EXP.GNFS.ZS] to exports
d.rename(columns={'Exports of goods and services (% of GDP) [NE.EXP.GNFS.ZS]': 'exports'}, inplace=True)

# Imports of goods and services (% of GDP) [NE.IMP.GNFS.ZS] to imports
d.rename(columns={'Imports of goods and services (% of GDP) [NE.IMP.GNFS.ZS]': 'imports'}, inplace=True)

# Foreign direct investment, net inflows (% of GDP) [BX.KLT.DINV.WD.GD.ZS] to fdi
d.rename(columns={'Foreign direct investment, net inflows (% of GDP) [BX.KLT.DINV.WD.GD.ZS]': 'fdi'}, inplace=True)

# Trade (% of GDP) [NE.TRD.GNFS.ZS] to trade
d.rename(columns={'Trade (% of GDP) [NE.TRD.GNFS.ZS]': 'trade'}, inplace=True)

# GDP growth (annual %) [NY.GDP.MKTP.KD.ZG] to gdp_grow
d.rename(columns={'GDP growth (annual %) [NY.GDP.MKTP.KD.ZG]': 'gdp_grow'}, inplace=True)

# External balance on goods and services (% of GDP) [NE.RSB.GNFS.ZS] to trade_balance
d.rename(columns={'External balance on goods and services (% of GDP) [NE.RSB.GNFS.ZS]': 'balance'}, inplace=True)



In [7]:
d.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1519 entries, 0 to 1518
Data columns (total 10 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   country    1519 non-null   object
 1   code       1519 non-null   object
 2   year       1519 non-null   int64 
 3   Time Code  1519 non-null   object
 4   exports    1519 non-null   object
 5   imports    1519 non-null   object
 6   fdi        1519 non-null   object
 7   trade      1519 non-null   object
 8   gdp_grow   1519 non-null   object
 9   balance    1519 non-null   object
dtypes: int64(1), object(9)
memory usage: 118.8+ KB


In [8]:
# Duplicate the dataset - Se hace porque se va a borrar cosas y para no tener problemas
d2 = d

In [9]:
# Create a new colum with yeas as str
d2['year_str'] = d2['year'].astype(str)
d2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1519 entries, 0 to 1518
Data columns (total 11 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   country    1519 non-null   object
 1   code       1519 non-null   object
 2   year       1519 non-null   int64 
 3   Time Code  1519 non-null   object
 4   exports    1519 non-null   object
 5   imports    1519 non-null   object
 6   fdi        1519 non-null   object
 7   trade      1519 non-null   object
 8   gdp_grow   1519 non-null   object
 9   balance    1519 non-null   object
 10  year_str   1519 non-null   object
dtypes: int64(1), object(10)
memory usage: 130.7+ KB


## 3.2. Delete columns

In [10]:
d2 = d2.drop(columns=['Time Code']) # se corre una vez, no dos o más veces porque da error.
d2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1519 entries, 0 to 1518
Data columns (total 10 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   country   1519 non-null   object
 1   code      1519 non-null   object
 2   year      1519 non-null   int64 
 3   exports   1519 non-null   object
 4   imports   1519 non-null   object
 5   fdi       1519 non-null   object
 6   trade     1519 non-null   object
 7   gdp_grow  1519 non-null   object
 8   balance   1519 non-null   object
 9   year_str  1519 non-null   object
dtypes: int64(1), object(9)
memory usage: 118.8+ KB


## 3.3. Delete NaN

In [11]:
# Replace by NaN
d2 = d2.replace('..', np.nan)
d2.head()

country code  year    exports    imports       fdi      trade  \
0  Afghanistan  AFG  2017        NaN        NaN  0.274797        NaN   
1  Afghanistan  AFG  2018        NaN        NaN  0.661572        NaN   
2  Afghanistan  AFG  2019        NaN        NaN  0.124496        NaN   
3  Afghanistan  AFG  2020  10.420817  36.289077  0.064994  46.709895   
4  Afghanistan  AFG  2021  14.342153  37.069564  0.144401  51.411716   

    gdp_grow    balance year_str  
0   2.647003        NaN     2017  
1   1.189228        NaN     2018  
2   3.911603        NaN     2019  
3  -2.351101 -25.868260     2020  
4 -20.738839 -22.727411     2021

In [12]:
d2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1519 entries, 0 to 1518
Data columns (total 10 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   country   1519 non-null   object 
 1   code      1519 non-null   object 
 2   year      1519 non-null   int64  
 3   exports   1220 non-null   float64
 4   imports   1220 non-null   float64
 5   fdi       1150 non-null   float64
 6   trade     1220 non-null   float64
 7   gdp_grow  1433 non-null   float64
 8   balance   1220 non-null   float64
 9   year_str  1519 non-null   object 
dtypes: float64(6), int64(1), object(3)
memory usage: 118.8+ KB


In [13]:
# Drop NaN
d2 = d2.dropna() #cogera toda la linea de NaN y la borra, por lo que se reduce la base
d2.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1008 entries, 3 to 1517
Data columns (total 10 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   country   1008 non-null   object 
 1   code      1008 non-null   object 
 2   year      1008 non-null   int64  
 3   exports   1008 non-null   float64
 4   imports   1008 non-null   float64
 5   fdi       1008 non-null   float64
 6   trade     1008 non-null   float64
 7   gdp_grow  1008 non-null   float64
 8   balance   1008 non-null   float64
 9   year_str  1008 non-null   object 
dtypes: float64(6), int64(1), object(3)
memory usage: 86.6+ KB


# **4. Indexing**
Le decimos a python que no solo es una información puesta, sino que hay unas relaciones.

Categories of the dataset

In [15]:
# Indexing
d2 = d2.set_index(['year', 'country'])
d2.head()

code    exports    imports       fdi      trade   gdp_grow  \
year country                                                                  
2020 Afghanistan  AFG  10.420817  36.289077  0.064994  46.709895  -2.351101   
2021 Afghanistan  AFG  14.342153  37.069564  0.144401  51.411716 -20.738839   
2017 Albania      ALB  31.569798  46.624451  7.855448  78.194250   3.802599   
2018 Albania      ALB  31.571954  45.236230  7.946356  76.808184   4.019346   
2019 Albania      ALB  31.304270  44.974925  7.797920  76.279195   2.087712   

                    balance year_str  
year country                          
2020 Afghanistan -25.868260     2020  
2021 Afghanistan -22.727411     2021  
2017 Albania     -15.054653     2017  
2018 Albania     -13.664275     2018  
2019 Albania     -13.670655     2019

In [16]:
d2.info() #pais y año no salen como columnas que se puedan seleccionar, porque estan arriba. Entre más categodias se tengan para indexar, mucho mejor. Entre más datos indexados, mejores graficos se pueden realizar.

<class 'pandas.core.frame.DataFrame'>
MultiIndex: 1008 entries, (2020, 'Afghanistan') to (2022, 'Zimbabwe')
Data columns (total 8 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   code      1008 non-null   object 
 1   exports   1008 non-null   float64
 2   imports   1008 non-null   float64
 3   fdi       1008 non-null   float64
 4   trade     1008 non-null   float64
 5   gdp_grow  1008 non-null   float64
 6   balance   1008 non-null   float64
 7   year_str  1008 non-null   object 
dtypes: float64(6), object(2)
memory usage: 71.7+ KB


# **5. Descriptive statistics**

In [17]:
# Basic
d2.describe() #solo saca la estadistica descriptiva de las variables que son número

exports      imports          fdi        trade     gdp_grow  \
count  1008.000000  1008.000000  1008.000000  1008.000000  1008.000000   
mean     42.541756    47.748240     3.563746    90.289996     2.586084   
std      32.247551    28.042078    18.800117    58.115182     7.395210   
min       1.571162     1.127673  -394.471644     2.698834   -54.335876   
25%      22.775003    29.067187     1.069240    55.699663     0.884286   
50%      35.084904    41.640886     2.557996    77.078758     3.154638   
75%      51.147141    58.787271     4.511036   105.564017     5.299660   
max     213.222679   198.445418   234.248695   402.459681   153.492641   

           balance  
count  1008.000000  
mean     -5.206485  
std      16.587675  
min     -86.617105  
25%     -12.815523  
50%      -2.714714  
75%       2.902563  
max      50.709722

In [18]:
d2.describe().apply(lambda s: s.apply('{0:.2f}'.format)) #.2 es que se le esta pidiendo 2 decimales

exports  imports      fdi    trade gdp_grow  balance
count  1008.00  1008.00  1008.00  1008.00  1008.00  1008.00
mean     42.54    47.75     3.56    90.29     2.59    -5.21
std      32.25    28.04    18.80    58.12     7.40    16.59
min       1.57     1.13  -394.47     2.70   -54.34   -86.62
25%      22.78    29.07     1.07    55.70     0.88   -12.82
50%      35.08    41.64     2.56    77.08     3.15    -2.71
75%      51.15    58.79     4.51   105.56     5.30     2.90
max     213.22   198.45   234.25   402.46   153.49    50.71

In [19]:
d2.describe().apply(lambda s: s.apply('{:,.2f}'.format)) #con una coma.

exports   imports       fdi     trade  gdp_grow   balance
count  1,008.00  1,008.00  1,008.00  1,008.00  1,008.00  1,008.00
mean      42.54     47.75      3.56     90.29      2.59     -5.21
std       32.25     28.04     18.80     58.12      7.40     16.59
min        1.57      1.13   -394.47      2.70    -54.34    -86.62
25%       22.78     29.07      1.07     55.70      0.88    -12.82
50%       35.08     41.64      2.56     77.08      3.15     -2.71
75%       51.15     58.79      4.51    105.56      5.30      2.90
max      213.22    198.45    234.25    402.46    153.49     50.71

#### Miercoles 16 de octubre se hace el taller - 20%.
#### Lunes 21 de octubre parcial - 15%.

Explicar en ingles que es lo que se hizo en el archivo con la base y para cerrar, analizar los datos contextualizando, diciendo más o menos. Desviación estandar.
(no más de 5 minutos)